Make Julia exceptions wrapped in Python Exceptions more informative

galenlynch opened the pull request:
I have added a back-trace to the description of exceptions coming from Julia
that are subsequently wrapped in Python exceptions.

This makes it much easier to find the source of errors in Julia code being
called from Python functions.

Closes #607

tkf commented:
Julia's backtrace can be quite long sometimes so I'm not sure constructing a string all the time is a good idea.  I think the right approach is to wrap exception lazily and then generate the exception message in Python side on demand:

```python
class PyCallException(Exception):
    def __init__(self, exception, backtrace):
        self.exception = exception
        self.backtrace = backtrace

   def __str__(self):
       return f"Julia exception: {showerror_string(self.exception)}"

   def print_backtrace(self):
       print(sprint(showerror, self.exception, self.backtrace))


class PyCallKeyError(PyCallException, KeyError):
   ...
```

where `print_backtrace` is a non-standard method you need to call manually (e.g., via a debugger or logger).

One thing I'm not sure about my approach is the lifetime of the backtrace returned by `catch_backtrace()`.  It's just a `Vector{Union{Ptr{Nothing}, InterpreterIP}}` and I wonder how long those pointers are valid.

Another much more simple-minded approach is to just put 

```julia
@debug "Throwing error in Python" exception=(e, bt)
```

in `pyraise(e)`.  This way, you can just `ENV["JULIA_DEBUG"] = "PyCall"` to look at the stacktraces.

tkf commented:
I didn't mean that there is a length limit for the string generated from Python exceptions.  I'm not sure if such limit exists (other than RAM).

> generating a string would a small performance hit compared to exception handling

When the exception includes objects with a lot of type parameters (like when using ForwardDiff.jl), I find that stacktrace to be huge and very slow to print everything.  I don't think you'd want to pay that cost simply because a Julia function throws (say) a `KeyError`.  Furthermore, many Python APIs use exception type as the part of the interface (at least more than Julia, I think).  So, the exception message may be ignored and generating it can be totally wasteful.  That's why I don't think converting stacktrace to a string all the time is optimal.

> I find the error messages so terse that they're seldom helpful.

Yeah, I want to improve it, too.

> Is there some other mechanism to include the Julia stack in the python exception?

AFIK `Exception.__str__` is the only generic way to do this.  Actually, including stacktrace in it (like you do in this PR) is not so bad since it means that it is likely that Python is printing its own stacktrace.

galenlynch commented:
Thanks for thinking through this in so much detail!

If I understand you correctly, your solution requires making a whole bunch of new exception classes in python and putting the Julia stacktrace in the `__str__` attribute. While this may be the right way of solving the problem, I think it's beyond the scope of what I need and I won't implement it in this PR.

Since your suggestion of adding `@debug "Throwing error in Python" exception=(e, bt)` in `pyraise(e)` would satisfy my need of determining where exceptions in Julia are coming from when I am developing some code, then I would be happy to implement it and see how it works as part of this PR. If not I'm happy to close the PR.

stevengj commented on src/exception.jl:
```diff
@@ -132,7 +132,7 @@ be called via Python C-API, it tries to not throw at all cost.
     try
         io = IOBuffer()
-        showerror(io, e)
+        showerror(io, e, catch_backtrace())
```
I'm worried that this isn't valid  We should really only call `catch_backtrace()` in a `catch` block, otherwise its value may be unpredictable.

Can the caller to `showerror_string` pass in the backtrace if it is available?

galenlynch commented:
@stevengj I tried to address your concern, though the solution I came up with isn't the prettiest.

I have changed `pyraise` to accept an optional backtrace, which will be passed to `showerror_string`. Since `pyraise` only calls `showerror_string` for non-`PyError` exceptions, the `pyraise` method accepting `PyError`s has a dummy argument so backtraces can be passed to either method, though they will ultimately be ignored if they are paired with a `PyError` exception.

I also added `catch_backtrace()` to all calls of `pyraise` inside a try-catch block in PyCall.jl.

I think these changes result in more verbose and informative error messages from exceptions originating in Julia code, while not calling `catch_backtrace` in inappropriate contexts. Additionally, if users want to supply a backtrace to `pyraise`, they will now be able to do so.

I understand that these changes obscure the flow of logic, so I would totally understand if you don't like them. However, my goal has already been achieved with the original commit, which I agree is not a general solution as `pyraise` is exported and can be called outside of a try-catch block where there may not be a current exception. As such, I might just tweak PyCall's code locally to show a stack trace when I need one, instead of developing a more generally robust solution than the one I've added here.

stevengj commented on src/callback.jl:
```diff
@@ -39,7 +39,7 @@ function _pyjlwrap_call(f, args_::PyPtr, kw_::PyPtr)
         return pyreturn(ret)
     catch e
-        pyraise(e)
+        pyraise(e, catch_backtrace())
```
Might be cleaner to have a `@pyraise e` macro that inserted the `catch_backtrace()` call:
```jl
macro pyraise(e)
    :(pyraise($(esc(e)), catch_backtrace())
end
```

galenlynch replied:
Done

stevengj commented:
Overall, this looks good to me.

tkf commented:
@stevengj I'm a bit worried about the overhead for this https://github.com/JuliaPy/PyCall.jl/pull/608#issuecomment-434915444, since exceptions in Python are considered "lightweight" and rendering backtrace can be time-consuming in Julia sometimes.  But we can test if that's the case in the wild by just releasing it... :smiling_imp:

galenlynch commented:
Given the choice between fast exceptions that are hard to locate, and slower exceptions that tell you exactly where the problem is, I would personally choose the latter. For my use case, hunting down the exception would have been nearly impossible without a backtrace. However, @tkf is right that converting the backtrace into a string takes time:

```julia
julia> bt = backtrace()
16-element Array{Union{Ptr{Nothing}, InterpreterIP},1}:
 Ptr{Nothing} @0x00007f0af9dcb176                                                                                                                                                                         
 Ptr{Nothing} @0x00007f0af9f34800                                                                                                                                                                         
 Ptr{Nothing} @0x00007f0af9f34529                                                                                                                                                                         
 Ptr{Nothing} @0x00007f0af9f34eac                                                                                                                                                                         
 Ptr{Nothing} @0x00007f0af9f3562f                                                                                                                                                                         
 Base.InterpreterIP(CodeInfo(
 1 ─ %1 = backtrace()                                                       │
 │        bt = %1                                                           │
 └──      return %1                                                         │
), 0x0000000000000000)
 Ptr{Nothing} @0x00007f0af9f360ad                                                                                                                                                                         
 Ptr{Nothing} @0x00007f0af9e012fc                                                                                                                                                                         
 Ptr{Nothing} @0x00007f0af9ddc590                                                                                                                                                                         
 Ptr{Nothing} @0x00007f0ae16ea992                                                                                                                                                                         
 Ptr{Nothing} @0x00007f0af9dcb176                                                                                                                                                                         
 Ptr{Nothing} @0x00007f0ae1d59ac3                                                                                                                                                                         
 Ptr{Nothing} @0x00007f0ae1d59d24                                                                                                                                                                         
 Ptr{Nothing} @0x00007f0af9dcb176                                                                                                                                                                         
 Ptr{Nothing} @0x00007f0af9de6afb                                                                                                                                                                         
 Ptr{Nothing} @0x0000000000000000

julia> @benchmark PyCall.showerror_string(ErrorException("foo"))
BenchmarkTools.Trial: 
  memory estimate:  336 bytes
  allocs estimate:  8
  --------------
  minimum time:     339.516 ns (0.00% GC)
  median time:      345.489 ns (0.00% GC)
  mean time:        399.394 ns (6.74% GC)
  maximum time:     11.166 μs (93.70% GC)
  --------------
  samples:          10000
  evals/sample:     219

julia> @benchmark PyCall.showerror_string(ErrorException("foo"), bt)
BenchmarkTools.Trial: 
  memory estimate:  23.47 KiB
  allocs estimate:  519
  --------------
  minimum time:     9.299 ms (0.00% GC)
  median time:      9.652 ms (0.00% GC)
  mean time:        9.734 ms (0.00% GC)
  maximum time:     14.666 ms (0.00% GC)
  --------------
  samples:          514
  evals/sample:     1
```
I don't have a sense for how common it is for python code to rely on exception handling for non-exceptional situations, which would be necessary to determine the real world performance impact of this PR.

Merging this PR does not preclude a future lazy implementation, like @tkf suggested. Right now, passing a back trace to `pyraise` is optional, and calls to `pyraise` within PyCall.jl supply that back trace. If and when a better implementation is made, then it would be trivial to change the default behaviour within PyCall.jl.

tkf commented:
I was rather talking about the case that printing backtrace takes a few tens of seconds.  I remember it happens in case there are a lot of type parameters.  I couldn't find the example now but here is a rather contrived example that takes about a half second:

```julia
using ForwardDiff

nthderiv(f, n, x = 0.0) =
    if n == 1
        ForwardDiff.derivative(f, x)
    else
        ForwardDiff.derivative(z -> nthderiv(f, n - 1, z), x)
    end

for n in 1:100
    f = Symbol("f$n")
    g = Symbol("f$(n-1)")
    @eval @noinline $f(x, bang) = $g(x, bang) + 1
end
function f0(n, bang)
    bang && error("bang!")
    return 0
end

ex = try
    nthderiv(x -> f100(x, true), 10)
catch e
    e, catch_backtrace()
end
```

```julia
julia> sizeof(@time sprint(showerror, ex...)) / 2^20
  0.465707 seconds (3.96 M allocations: 225.571 MiB, 6.76% gc time)
11.887667655944824
```


But this generates a 11 MiB string so maybe there is no wonder why it is slow.  Actually, this is much faster than printing it to the terminal (via `tmux`) which takes more than a minute (at which time I terminated the process).  So maybe the slowness I remembered was due to the bottleneck in the terminal.  (Although I remember that sometimes it looked like Julia was JIT'ing something)

tkf commented:
FYI Python manual says:

> EAFP
> Easier to ask for forgiveness than permission. This **common Python coding style** assumes the existence of valid keys or attributes and catches exceptions if the assumption proves false.
> --- https://docs.python.org/3/glossary.html#term-eafp

stevengj commented:
It seems like showerror should be fixed in Base to truncate the backtrace if it is longer than a couple kB. 

I’m not so concerned about performance here at the moment, though. I doubt that people will go to the trouble of calling Julia code from Python except for fairly expensive functions—not for cheap functions like dictionary lookups where the cost of constructing a backtrace is a concern.
